File: sequencial.c
#include <stdio.h>
#include <stdlib.h>
#include <time.h>

int main(int argc, char *argv[]) {
    int l1, c1, l2, c2;

    // abre os arquivos
    FILE *matriz1 = fopen(argv[1], "r");
    FILE *matriz2 = fopen(argv[2], "r");

    // le o tamanho
    fscanf(matriz1, "%d %d", &l1, &c1);
    fscanf(matriz2, "%d %d", &l2, &c2);

    // aloca matriz
    int **new_matriz1 = (int **) malloc(l1 * sizeof(int *));
    int **new_matriz2 = (int **) malloc(l2 * sizeof(int *));
    int **resultado = (int **) malloc(l1 * sizeof(int *));

    for (int i = 0; i < l1; i++) {
        new_matriz1[i] = (int *) malloc(c1 * sizeof(int));
        resultado[i] = (int *) malloc(c2 * sizeof(int));
    }

    for (int i = 0; i < l2; i++) {
        new_matriz2[i] = (int *) malloc(c2 * sizeof(int));
    }

    // leitura das matrizes
    for (int i = 0; i < l1; i++) {
        for (int j = 0; j < c1; j++) {
            fscanf(matriz1, "%d", &new_matriz1[i][j]);
        }
    }

    for (int i = 0; i < l2; i++) {
        for (int j = 0; j < c2; j++) {
            fscanf(matriz2, "%d", &new_matriz2[i][j]);
        }
    }

    // multiplicação
    clock_t inicia = clock();
    for (int i = 0; i < l1; i++) {
        for (int j = 0; j < c2; j++) {
            resultado[i][j] = 0;
            for (int k = 0; k < c1; k++) {
                resultado[i][j] += new_matriz1[i][k] * new_matriz2[k][j];
            }
        }
    }
    clock_t fim = clock();

    // cálculo de tempo
    double tempo = (double)(fim - inicia) / CLOCKS_PER_SEC;

    // gravação do resultado em outro arquivo
    FILE *saida = fopen("saida.txt", "w");
    fprintf(saida, "%d %d\n", l1, c2);  // tamanho da matriz resultante

    for (int i = 0; i < l1; i++) {
        for (int j = 0; j < c2; j++) {
            fprintf(saida, "%d %d %d\n", i, j, resultado[i][j]);  // posição linha, coluna e valor
        }
    }

    fprintf(saida, "%.6f", tempo);  // tempo de execução
    fclose(saida);
	
return 0;
}

File: sequencial.py
import sys
import time

# Função para ler uma matriz a partir de um arquivo
def le_matriz_arquivo(filename):
    with open(filename, 'r') as f:
        # Lê o tamanho da matriz
        m, n = map(int, f.readline().split())

        # Inicializa a matriz com zeros
        matriz = [[0 for _ in range(n)] for _ in range(m)]

        # Lê os valores da matriz
        for i in range(m):
            linha = f.readline().split()
            for j in range(n):
                matriz[i][j] = int(linha[j])

    return matriz


# Função para multiplicar duas matrizes
def multiplica_matrizes(matriz1, matriz2):
    # Verifica se as matrizes são compatíveis para a multiplicação
    if len(matriz1[0]) != len(matriz2):
        return None

    # Inicializa a matriz resultante com zeros
    saida = [[0 for _ in range(len(matriz2[0]))] for _ in range(len(matriz1))]

    # Realiza a multiplicação
    for i in range(len(matriz1)):
        for j in range(len(matriz2[0])):
            for k in range(len(matriz2)):
                saida[i][j] += matriz1[i][k] * matriz2[k][j]

    return saida

# Lê as matrizes de entrada
m1_filename = sys.argv[1]
m2_filename = sys.argv[2]
m1 = le_matriz_arquivo(m1_filename)
m2 = le_matriz_arquivo(m2_filename)

# Multiplica as matrizes
tempo_inicio = time.time()
resultado = multiplica_matrizes(m1, m2)
tempo_fim = time.time()

# Escreve o resultado no arquivo de saída
output_filename = sys.argv[3]
with open(output_filename, 'w') as f:
    # Escreve o tamanho da matriz
    f.write(f"{len(resultado)} {len(resultado[0])}\n")

    # Escreve os valores da matriz
    for i in range(len(resultado)):
        for j in range(len(resultado[0])):
            if resultado[i][j] != 0:
                f.write(f"c{i+1}{j+1} {resultado[i][j]}\n")

    # Escreve o tempo de execução
    f.write(f"{tempo_fim - tempo_inicio:.2f}")
